planex/cmd/clone.py

```diff
@@ -122,6 +141,109 @@ def apply_patchqueue(base_repo, pq_repo, pq_dir):
+           for (url2, commitish2) in gathered
+           if url1 == url2):
+        sys.exit("error: cloning two git repositories with the same "
+                 "name but different commitish is not supported.")
+
+    for url, commitish in gathered:
+        print('echo "Cloning %s"' % url)
+        if args.jenkins:
+            clone_jenkins(url, args.repos, commitish, args.credentials)
+        # clone is assumed for all other flags
+        else:
+            util.makedirs(args.repos)
+            try:
+                nodetached = args.patchqueue or args.repatched
+                clone(url, args.repos, commitish, nodetached)
+            except git.GitCommandError as gce:
+                print(gce.stderr)
+
+
+def assemble_patchqueue(args, pin):
+    """
+    Assemble patchqueues using Source0 and PatchQueue0 using
+    the active branches in the cloned sources.
+    """
+    sources = {
+        key: src['URL']
+        for key, src in pin.sources.items()
+        if src.get('commitish', False)
+    }
+    patchqueues = {
+        key: (pq['URL'], dirname(pq['prefix']))
+        for key, pq in pin.patchqueue_sources.items()
+        if pq.get('commitish', False)
+    }
+
+    if 0 not in sources:
+        sys.exit("error: planex-clone requires Source0 to point to "
+                 "a git repository.")
+    if 0 not in patchqueues:
+        sys.exit("error: planex-clone requires PatchQueue0 to point "
+                 "to a git repository.")
+
+    if len(patchqueues.keys()) > 1:
+        sys.exit(
+            "error: planex-clone does not support the cloning and "
+            "assembly of multiple sources and patchqueues, currently "
+            "this case needs to be handled by hands.")
+    try:
+        src_url = sources[0]
+        pq_url, pq_prefix = patchqueues[0]
+        src_repo = git.Repo(join(args.repos, repo_name(src_url)))
+        pq_repo = git.Repo(join(args.repos, repo_name(pq_url)))
+        apply_patchqueue(src_repo, pq_repo, pq_prefix)
+        print("By default the PIN file will instruct `planex` to "
+              "`git archive` the source repository and the patchqueue "
+              "repository at the commitish specified in the pin file, "
+              "then assemble the patchqueue when building the RPM.\n"
+              "If you want to build using the HEAD of the source repository, "
+              "where you have all your ongoing modifications applied, you "
+              "can edit the pin file as follows: \n"
+              "\t- delete the `PatchQueue0` section, and\n"
+              "\t- set the correct commitish for `Source0` (usually the name "
+              "of the active branch).")
+    except git.GitCommandError as gce:
+        print(gce.stderr)
+
+
+def assemble_repatched(args, pin):
+    """
+    Assemble an upstream centos package and apply the upstream patchqueue,
+    then apply the local patchqueue.
+    """
+    print(args, pin)
+    raise Exception("Unimplemented")
```

Comment:
Use the standard exception `NotImplementedError`

Reply:
Thanks!!! I could not remember or find it